### ISSUE TITLE
Type alias circular reference is not detected

### ISSUE THREAD

**Issue by osdm** [Microsoft/TypeScript#2980]
According to spec, circular references are prohibited for type aliases and should be reported as errors. But in 1.4 it depends on the order of declarations. For example, no error is reported in the following code:
```ts
type typeAlias = typeof varOfAliasedType;
var varOfAliasedType: typeAlias;
```
Error is reported only if order of statements is reversed.

One more example:
```ts
var varOfAliasedType = func();
function func(): typeAlias { return null; }
type typeAlias = typeof varOfAliasedType;
```
Error is reported only if order of the first two statements is reversed.

**Comment by ahejlsberg** [comment]
This is bit of an odd one. We've always had the rule that if a variable references itself (directly or indirectly) in its own declaration, we give it type `any`, but only issue an error in `-noImplicitAny` mode. As I recall this rule was introduced to allow JavaScript code such as the following (somewhat suspect, but technically correct) code to compile without errors:
```typescript
var x = x + 1;
```
We currently use the same rule even when a variable has a type annotation:
```typescript
var y: typeof y;  // y is given type any without error
```
That behavior doesn't really make sense. When a variable has a type annotation, clearly the intent is for it to have a *valid* type and the above should be an error. As your bug report indicates, it *sometimes* is, depending on which construct we check first. But we should fix it to always be an error.
